src/Package/Impl/ProjectSystem/Commands/SendFileCommandBase.cs

```diff
@@ -0,0 +1,101 @@
+            var outputWindow = workflow.ActiveWindow.InteractiveWindow;
+            uint cookie = 0;
+            try {
+                var session = workflow.RSession;
+                statusBar.SetText(Resources.Info_CompressingFiles);
+                statusBar.Progress(ref cookie, 1, "", 0, 0);
+
+                int count = 0;
+                uint total = (uint)files.Count() * 2; // for compressing and sending
+                string compressedFilePath = string.Empty;
+                await Task.Run(() => {
+                    compressedFilePath = _fs.CompressFiles(files, projectDir, new Progress<string>((p) => {
+                        Interlocked.Increment(ref count);
+                        statusBar.Progress(ref cookie, 1, string.Format(Resources.Info_CompressingFile, Path.GetFileName(p)), (uint)count, total);
+                        string dest = p.MakeRelativePath(projectDir).ProjectRelativePathToRemoteProjectPath(remotePath, projectName);
+                        _coreShell.DispatchOnUIThread(() => {
+                            outputWindow.WriteLine(string.Format(Resources.Info_LocalFilePath, p));
+                            outputWindow.WriteLine(string.Format(Resources.Info_RemoteFilePath, dest));
+                        });
+                    }), CancellationToken.None);
+                    statusBar.Progress(ref cookie, 0, "", 0, 0);
+                });
+
+                using (var fts = new DataTransferSession(session, _fs)) {
+                    cookie = 0;
+                    statusBar.SetText(Resources.Info_TransferringFiles);
+
+                    total = (uint)_fs.FileSize(compressedFilePath);
+
+                    var remoteFile = await fts.SendFileAsync(compressedFilePath, true, new Progress<long>((b) => {
+                        statusBar.Progress(ref cookie, 1, Resources.Info_TransferringFiles, (uint)b, total);
+                    }));
+
+                    statusBar.SetText(Resources.Info_ExtractingFilesInRHost);
+                    await session.EvaluateAsync<string>($"rtvs:::save_to_project_folder({remoteFile.Id}, {projectName.ToRStringLiteral()}, '{remotePath.ToRPath()}')", REvaluationKind.Normal);
+
+                    _coreShell.DispatchOnUIThread(() => {
+                        outputWindow.WriteLine(Resources.Info_TransferringFilesDone);
+                    });
+                }
+            } catch (IOException ex) {
+                _coreShell.ShowErrorMessage(string.Format(CultureInfo.InvariantCulture, Resources.Error_CannotTransferFile, ex.Message));
+            } catch (RHostDisconnectedException) {
+                _coreShell.DispatchOnUIThread(() => {
+                    outputWindow.WriteErrorLine(Resources.Error_CannotTransferNoRSession);
+                });
+            } finally {
+                statusBar.Progress(ref cookie, 0, "", 0, 0);
+                statusBar.SetText(currentStatusText);
+                await _coreShell.SwitchToMainThreadAsync();
```

Comment:
Which is why I prefer to extract code chunk so that `SwitchToBackgroundThread` is called at the very beginning of the method.

---
In reply to: [81013007](https://github.com/Microsoft/RTVS/pull/2405#discussion_r81013007) [](ancestors = 81013007)